src/command/dev/protocol/leader.rs

```diff
@@ -267,6 +269,23 @@ impl LeaderSession {
 
         if let Vacant(e) = self.subgraphs.entry((name.to_string(), url.clone())) {
             e.insert(sdl.to_string());
+
+            // Followers add subgraphs, but sometimes those subgraphs depend on each other
+            // (e.g., through extending a type in another subgraph). When that happens,
+            // composition fails until _all_ subgraphs are loaded in. This acknowledges the
+            // follower's message when we haven't loaded in all the subgraphs, deferring
+            // composition until we have at least the number of subgraphs represented in the
+            // supergraph.yaml file
+            //
+            // This applies only when the supergraph.yaml file is present. Without it, we will
+            // try composition each time we add a subgraph
+            if let Some(supergraph_config) = self.supergraph_config.clone() {
```

Comment:
How expensive could the clone be here? I think we're doing this to get around the ownership problem of taking from the `Option`, but I'm wondering if there's a `len()` method on `self.supergraph_config` so we don't need to transform it into an iterator?

Reply:
that's a good question, and kind of a tricky one. `len()` only works on iterables; so, there's no `len()` off of `SupergraphConfig`, and there's also no way to get the subgraphs out of that config without having to do error handling (the return type of `get_subgraph_definitions()` is `ConfigResult<Vec<SubgraphDefinition>>`); otherwise, we might have been able to do something like `supergraph_config.subgraphs.len()`.

So, to get an iterable to do `len()` for our comparison, when limited to just `into_iter()`, we need an owned value. That's why we `clone()` here.

Cost is a good thing to call out, though! I think for very large supergraph configs, this might be slow; the reason I'm not too worried about it yet (which is to say, why I'm leaving it as some later iteration) is because for large supergraph configs, the noise generated by intermediate composition errors is more annoying and misleading than a little bit of added time to add subgraphs to the internal representation of the config

Reply:
This seems like a failing of `SupergraphConfig` in `apollo-federation-types`. Adding a getter for this data would probably be a good idea.

Reply:
there is, `get_subgraph_definitions()`, but internally it composes the supergraph before giving back the subgraphs; so, it puts the cart a bit before the horse from our perspective, in that we're _also_ composing

a more permissive getter that sends back whatever it has without composition would be nice